File: lexer/go/lexer/lexer.go
package lexer

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func isWhitespace(ch byte) bool {
	return ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n'
}

type Lexer struct {
	input string

	// Reader state
	ch           byte // Current read character
	la           byte // Lookahead character (ie. next character in input)
	readPosition int
	readLineNo   uint
	readColNo    uint

	// Current position
	position int
	lineNo   uint
	colNo    uint
}

func New(input string) *Lexer {
	l := &Lexer{input: input, readLineNo: 1, readColNo: 1}
	return l
}

func (l *Lexer) NextToken() Token {
	l.position = l.readPosition
	l.lineNo = l.readLineNo
	l.colNo = l.readColNo
	l.readChar()

	switch l.ch {
	case '(':
		return l.match(Lparen)
	case ')':
		return l.match(Rparen)
	case '[':
		return l.match(Lbracket)
	case ']':
		return l.match(Rbracket)
	case '{':
		return l.match(Lcurly)
	case '}':
		return l.match(Rcurly)
	case ';':
		return l.match(Semicolon)
	case ':':
		return l.match(Colon)
	case ',':
		return l.match(Comma)
	case '+':
		return l.matchIf('=', AssignPlus, Plus)
	case '-':
		return l.matchIf('=', AssignMinus, Minus)
	case '*':
		return l.matchIf('=', AssignMultiply, Multiply)
	case '/':
		return l.matchIf('=', AssignDivide, Divide)
	case '%':
		return l.matchIf('=', AssignMod, Mod)
	case '<':
		return l.matchIf('=', LessEqual, LessThan)
	case '>':
		return l.matchIf('=', GreaterEqual, GreaterThan)
	case '=':
		return l.matchIf('=', Equal, Assign)
	case '!':
		return l.matchIf('=', NotEqual, Illegal)
	case '~':
		return l.matchIf('=', AssignConcat, Concat)
	case '^':
		return l.match(Pow)
	case '"':
		return l.matchString()
	case '#':
		return l.matchComment()
	case 0:
		return l.match(Eof)
	default:
		if isLetter(l.ch) {
			return l.matchIdentifier()
		} else if isDigit(l.ch) {
			return l.matchNumber()
		} else if isWhitespace(l.ch) {
			return l.matchWhitespace()
		} else {
			return l.match(Illegal)
		}
	}
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
		if l.ch == '\n' {
			l.readLineNo += 1
			l.readColNo = 0
		}
		l.readColNo += 1
		l.readPosition += 1
	}
	if l.readPosition >= len(l.input) {
		l.la = 0
	} else {
		l.la = l.input[l.readPosition]
	}
}

func (l *Lexer) match(tokenType TokenType) Token {
	return Token{
		LineNo:  l.lineNo,
		ColNo:   l.colNo,
		Type:    tokenType,
		Literal: l.input[l.position:l.readPosition]}
}

func (l *Lexer) matchIf(
	test byte,
	typeIf TokenType,
	typeElse TokenType) Token {
	if l.la == test {
		l.readChar()
		return l.match(typeIf)
	} else {
		return l.match(typeElse)
	}
}

func (l *Lexer) matchIdentifier() Token {
	for isLetter(l.la) || isDigit(l.la) {
		l.readChar()
	}
	tok := l.match(Identifier)
	tok.Type = LookupIdenitifer(tok.Literal)
	return tok
}

func (l *Lexer) matchWhitespace() Token {
	for isWhitespace(l.la) {
		l.readChar()
	}
	return l.match(Whitespace)
}

func (l *Lexer) matchComment() Token {
	for l.ch != 0 && l.ch != '\n' {
		l.readChar()
	}
	return l.match(Comment)
}

func (l *Lexer) matchNumber() Token {
	for isDigit(l.la) || l.la == '_' {
		l.readChar()
	}
	if l.la == '.' {
		l.readChar()
		for isDigit(l.la) || l.la == '_' {
			l.readChar()
		}
		if l.la == 'e' || l.la == 'E' {
			l.readChar()
			if l.la == '-' || l.la == '+' {
				l.readChar()
			}
			for isDigit(l.la) || l.la == '_' {
				l.readChar()
			}
		}
		return l.match(Float)
	}
	return l.match(Integer)
}

func (l *Lexer) matchString() Token {
	l.readChar()
	for l.la != 0 && l.la != '"' {
		if l.la == '\\' {
			l.readChar()
		}
		l.readChar()
	}
	if l.la == '"' {
		l.readChar()
	}
	return l.match(String)
}

File: lexer/java_custom/src/main/java/com/cameronzemek/zscript/Lexer.java
package com.cameronzemek.zscript;

public class Lexer {
  private String input;
  private int length;

  private char ch;
  private char la;
  private int readPosition;
  private int readLineNo = 1;
  private int readColNo = 1;

  private int position;
  private int lineNo;
  private int colNo;

  private static boolean isLetter(char b) {
    return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_';
  }

  private static boolean isDigit(char b) { return b >= '0' && b <= '9'; }

  private static boolean isWhitespace(char b) {
    return b == ' ' || b == '\t' || b == '\r' || b == '\n';
  }

  public Lexer(String input) {
    this.input = input;
    this.length = input.length();
  }

  private void readChar() {
    if (readPosition >= length) {
      ch = 0;
    } else {
      ch = input.charAt(readPosition);
      if (ch == '\n') {
        readLineNo++;
        readColNo = 0;
      }
      readColNo++;
      readPosition++;
    }
    if (readPosition >= length) {
      la = 0;
    } else {
      la = input.charAt(readPosition);
    }
  }

  private Token match(TokenType type) {
    String text = input.substring(position, readPosition);
    return new Token(lineNo, colNo, type, text);
  }

  private Token matchIf(char ifCh, TokenType ifType, TokenType elseType) {
    if (la == ifCh) {
      readChar();
      return match(ifType);
    } else {
      return match(elseType);
    }
  }

  private Token matchString() {
    readChar(); // consume '"'
    while (ch != '"' && ch != 0) {
      if (ch == '\\') {
        readChar();
      }
      readChar();
    }
    return match(TokenType.STRING);
  }

  private Token matchComment() {
    while (ch != 0 && ch != '\n') {
      readChar();
    }
    return match(TokenType.COMMENT);
  }

  private Token matchWhitespace() {
    while (isWhitespace(la)) {
      readChar();
    }
    return match(TokenType.WHITESPACE);
  }

  private Token matchIdentifier() {
    while (isLetter(la) || isDigit(la)) {
      readChar();
    }
    String text = input.substring(position, readPosition);
    TokenType type = lookupKeyword(text);
    return new Token(lineNo, colNo, type, text);
  }

  private static TokenType lookupKeyword(String identifier) {
    switch (identifier) {
    case "function":
      return TokenType.KEYWORD_FUNCTION;
    case "return":
      return TokenType.KEYWORD_RETURN;
    case "true":
      return TokenType.KEYWORD_TRUE;
    case "false":
      return TokenType.KEYWORD_FALSE;
    case "and":
      return TokenType.KEYWORD_AND;
    case "or":
      return TokenType.KEYWORD_OR;
    case "xor":
      return TokenType.KEYWORD_XOR;
    case "not":
      return TokenType.KEYWORD_NOT;
    case "let":
      return TokenType.KEYWORD_LET;
    case "if":
      return TokenType.KEYWORD_IF;
    case "else":
      return TokenType.KEYWORD_ELSE;
    case "while":
      return TokenType.KEYWORD_WHILE;
    case "foreach":
      return TokenType.KEYWORD_FOREACH;
    case "as":
      return TokenType.KEYWORD_AS;
    default:
      return TokenType.IDENTIFIER;
    }
  }

  private Token matchNumber() {
    while (isDigit(la) || la == '_') {
      readChar();
    }
    if (la == '.') {
      readChar();
      while (isDigit(la) || la == '_') {
        readChar();
      }
      if (la == 'e' || la == 'E') {
        readChar();
        if (la == '-' || la == '+') {
          readChar();
        }
        while (isDigit(la) || la == '_') {
          readChar();
        }
      }
      return match(TokenType.FLOAT);
    } else {
      return match(TokenType.INTEGER);
    }
  }

  public Token getNext() {
    position = readPosition;
    lineNo = readLineNo;
    colNo = readColNo;
    readChar();

    switch (ch) {
    case 0:
      return match(TokenType.EOF);
    case '^':
      return match(TokenType.POW);
    case ';':
      return match(TokenType.SEMICOLON);
    case ':':
      return match(TokenType.COLON);
    case ',':
      return match(TokenType.COMMA);
    case '{':
      return match(TokenType.LCURLY);
    case '}':
      return match(TokenType.RCURLY);
    case '[':
      return match(TokenType.LBRACKET);
    case ']':
      return match(TokenType.RBRACKET);
    case '(':
      return match(TokenType.LPAREN);
    case ')':
      return match(TokenType.RPAREN);
    case '+':
      return matchIf('=', TokenType.ASSIGN_PLUS, TokenType.PLUS);
    case '-':
      return matchIf('=', TokenType.ASSIGN_MINUS, TokenType.MINUS);
    case '*':
      return matchIf('=', TokenType.ASSIGN_MULTIPLY, TokenType.MULTIPLY);
    case '/':
      return matchIf('=', TokenType.ASSIGN_DIVIDE, TokenType.DIVIDE);
    case '%':
      return matchIf('=', TokenType.ASSIGN_MOD, TokenType.MOD);
    case '~':
      return matchIf('=', TokenType.ASSIGN_CONCAT, TokenType.CONCAT);
    case '=':
      return matchIf('=', TokenType.EQUAL, TokenType.ASSIGN);
    case '>':
      return matchIf('=', TokenType.GREATER_EQUAL, TokenType.GREATER_THAN);
    case '<':
      return matchIf('=', TokenType.LESS_EQUAL, TokenType.LESS_THAN);
    case '!':
      return matchIf('=', TokenType.NOT_EQUAL, TokenType.ILLEGAL);
    case '"':
      return matchString();
    case '#':
      return matchComment();
    default:
      if (isWhitespace(ch)) {
        return matchWhitespace();
      } else if (isLetter(ch)) {
        return matchIdentifier();
      } else if (isDigit(ch)) {
        return matchNumber();
      }
      return match(TokenType.ILLEGAL);
    }
  }
}
